# PROJECT: WRANGLE AND ANALYZE DATA

### IMPORTING USEFUL LIBRARIES

In [1]:
import pandas as pd
import numpy as np
import json
import requests
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

##  Data Gathering

In [2]:
# importing the fist dataset manually which is in csv format
arc_data = pd.read_csv('twitter-archive-enhanced.csv')

In [3]:
# programaticaly  download the second data which is in tsv format
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
response = requests.get(url)

#saving the tsv file
with open('image-predictions.tsv', mode = 'wb') as file:
    file.write(response.content)

In [4]:
# importing the image data into pandas
image_data =pd.read_csv('image-predictions.tsv',sep ='\t')

In [5]:
# opening the json file and reading it line by line into pandas datadrame
df = []
with open('tweet-json.txt', 'r') as file:
    for line in file:
        data = json.loads(line)
        df.append(data)
tweets_data = pd.DataFrame(df,columns = ['id','retweet_count','favorite_count'])
        


## Assess

In [6]:
# to check the firt 5 entry in the dataframe
arc_data.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [7]:
# to check the basic information of the dataframe
arc_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

this dataframe has 2356 entries and 17 columns with different datatypes such as integer,float and object.some columns as missing values such as in_reply_to_status_id,in_reply_to_user_id,retweeted_status_id,retweeted_status_user_id,retweeted_status_timestamp and expanded_urls

In [8]:
# to check the statistical summary of the dataset
arc_data.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [9]:
image_data.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [10]:
image_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_nu

In [11]:
image_data.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

In [12]:
tweets_data.head()

id  retweet_count  favorite_count
0  892420643555336193           8853           39467
1  892177421306343426           6514           33819
2  891815181378084864           4328           25461
3  891689557279858688           8964           42908
4  891327558926688256           9774           41048

id  retweet_count  favorite_count
0  892420643555336193           8853           39467
1  892177421306343426           6514           33819
2  891815181378084864           4328           25461
3  891689557279858688           8964           42908
4  891327558926688256           9774           41048

In [13]:
tweets_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   id              2354 non-null   int64
 1   retweet_count   2354 non-null   int64
 2   favorite_count  2354 non-null   int64
dtypes: int64(3)
memory usage: 55.3 KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   id              2354 non-null   int64
 1   retweet_count   2354 non-null   int64
 2   favorite_count  2354 non-null   int64
dtypes: int64(3)
memory usage: 55.3 KB


## Tidyness issues
1. rating_numerator and rating_denominator column should and being combined together and stored as float
2. doggo,floofer, pupper,puppo column should be conbine in a single column 
3. the image data and the tweet data should have being combined together with the archive data


## Quality issues 
1. high amount of missing values in some colums suchn as; in_reply_to_status_id,in_reply_to_user_id,retweeted_status_id,retweeted_status_user_id,retweeted_status_timestamp
2. expanded_url column not really important
3. the empty values stored as none in name column should be replaced with NAN
4. timestamp datatype is stored as object instead of  datetime
5. in p1,p2,p3 columns,some description are with underscore while others are not
6. some Column headers are not discriptive enough
7. under the name column,some values appear in lower case and it seems not to be real name e.g a,one,an,getting,the,trip
8. presence ofoutlier in rating_proportion(i.e rating_numerator/rating_denominator),retweet_count and varorite favorite_count

## Clean

In [14]:
# duplicating the dataframes
arc_data_2 = arc_data.copy()
image_data_2 = image_data.copy()
tweets_data_2 = tweets_data.copy()


### Tidyness  Issues 1 ; rating_numeration and rating denominator should be calculated in same column and stored as float

### Define


#### The rating proportion will be calculated by dividing the numerator by denominated

### Code

In [15]:
# divide the numerator by denomination and save the column as rating_proportion
rating =arc_data_2.rating_numerator/arc_data_2.rating_denominator
arc_data_2['rating'] =rating

In [16]:
# drop the rating numerator and rating demominator column
arc_data_2.drop(columns=['rating_numerator','rating_denominator'],inplace = True)

### Test

In [17]:
arc_data_2.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls      name doggo floofer  \
0  https://twitter.com/dog_rates/status/892420643...   Phineas  None    None   
1  https://twitter.com/dog_rates/status/892177421...     Tilly  None    None   
2  https://twitter.com/dog_rates/status/891815181...    Archie  None    None   
3  https://twitter.com/dog_rates/status/891689557...     Darla  None    None   
4  https://twitter.com/dog_rates/status/891327558...  Franklin  None    None   

  pupper puppo  rating  
0   None  None     1.3  
1   None  None     1.3  
2   None  None     1.2  
3   None  None     1.3  
4   None  None     1.2

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls      name doggo floofer  \
0  https://twitter.com/dog_rates/status/892420643...   Phineas  None    None   
1  https://twitter.com/dog_rates/status/892177421...     Tilly  None    None   
2  https://twitter.com/dog_rates/status/891815181...    Archie  None    None   
3  https://twitter.com/dog_rates/status/891689557...     Darla  None    None   
4  https://twitter.com/dog_rates/status/891327558...  Franklin  None    None   

  pupper puppo  rating  
0   None  None     1.3  
1   None  None     1.3  
2   None  None     1.2  
3   None  None     1.3  
4   None  None     1.2

### Tidyness Issues 2 ; doggo,floofer, pupper,puppo column should be conbine in a single column

### Define

#### The dogo,floofer,pupper and puppo will be combine in a column

### Code

In [18]:
# this code merge the three dog column into 1
breed_type = ['doggo','floofer','pupper','puppo']

arc_data_2[breed_type] = arc_data_2[breed_type].replace('None', np.nan)
def join_all(x):
    return ', '.join(x.dropna().astype(str))

arc_data_2['breed_type'] = arc_data_2[breed_type].apply(join_all, axis=1)

arc_data_2['breed_type']= arc_data_2['breed_type'].replace('', np.nan)


In [19]:
# drop the columns that were use in merging but are no longer necessary
arc_data_2.drop(columns=['doggo','floofer','pupper','puppo'],inplace= True)

### Test

In [20]:
# this output the value count of the new merge column 
arc_data_2['breed_type'].value_counts()

pupper            245
doggo              83
puppo              29
doggo, pupper      12
floofer             9
doggo, floofer      1
doggo, puppo        1
Name: breed_type, dtype: int64

pupper            245
doggo              83
puppo              29
doggo, pupper      12
floofer             9
doggo, floofer      1
doggo, puppo        1
Name: breed_type, dtype: int64

In [21]:
arc_data_2.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls      name  rating  \
0  https://twitter.com/dog_rates/status/892420643...   Phineas     1.3   
1  https://twitter.com/dog_rates/status/892177421...     Tilly     1.3   
2  https://twitter.com/dog_rates/status/891815181...    Archie     1.2   
3  https://twitter.com/dog_rates/status/891689557...     Darla     1.3   
4  https://twitter.com/dog_rates/status/891327558...  Franklin     1.2   

  breed_type  
0        NaN  
1        NaN  
2        NaN  
3        NaN  
4        NaN

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls      name  rating  \
0  https://twitter.com/dog_rates/status/892420643...   Phineas     1.3   
1  https://twitter.com/dog_rates/status/892177421...     Tilly     1.3   
2  https://twitter.com/dog_rates/status/891815181...    Archie     1.2   
3  https://twitter.com/dog_rates/status/891689557...     Darla     1.3   
4  https://twitter.com/dog_rates/status/891327558...  Franklin     1.2   

  breed_type  
0        NaN  
1        NaN  
2        NaN  
3        NaN  
4        NaN

### Tidyness Issues 3 ; The image data and the tweet data should have being combined together with the archive data

### Define

#### The three dataset will be merge into 1 called the master Archive

### Code

In [22]:
# merge the archive and the tweets data into twitter_archieve
twitter_archive= pd.merge(left=arc_data_2, right=tweets_data, left_on='tweet_id', right_on='id', how='inner')

In [23]:
# twitter_archive merge with image data to get the master_data
master_archive= twitter_archive.merge(image_data, on='tweet_id', how='inner')

In [24]:
master_archive.drop(columns=['id'],inplace = True)

### Test

In [25]:
master_archive.head(2)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   

                                       expanded_urls  ... img_num         p1  \
0  https://twitter.com/dog_rates/status/892420643...  ...       1     orange   
1  https://twitter.com/dog_rates/status/892177421...  ...       1  Chihuahua   

    p1_conf  p1_dog        p2   p2_conf  p2_dog        p3   p3_conf  p3_dog  
0  0.097049   False     bagel  0.085851   False    banana  0.076110   False  
1  0.323581    True  Pekinese  0.090647    True  papillon  0.068957    True  

[2 rows x 26 columns]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   

                                       expanded_urls  ... img_num         p1  \
0  https://twitter.com/dog_rates/status/892420643...  ...       1     orange   
1  https://twitter.com/dog_rates/status/892177421...  ...       1  Chihuahua   

    p1_conf  p1_dog        p2   p2_conf  p2_dog        p3   p3_conf  p3_dog  
0  0.097049   False     bagel  0.085851   False    banana  0.076110   False  
1  0.323581    True  Pekinese  0.090647    True  papillon  0.068957    True  

[2 rows x 26 columns]

In [26]:
master_archive.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2073 entries, 0 to 2072
Data columns (total 26 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2073 non-null   int64  
 1   in_reply_to_status_id       23 non-null     float64
 2   in_reply_to_user_id         23 non-null     float64
 3   timestamp                   2073 non-null   object 
 4   source                      2073 non-null   object 
 5   text                        2073 non-null   object 
 6   retweeted_status_id         79 non-null     float64
 7   retweeted_status_user_id    79 non-null     float64
 8   retweeted_status_timestamp  79 non-null     object 
 9   expanded_urls               2073 non-null   object 
 10  name                        2073 non-null   object 
 11  rating                      2073 non-null   float64
 12  breed_type                  320 non-null    object 
 13  retweet_count               2073 

In [27]:
master_archive.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.073000e+03           2.300000e+01         2.300000e+01   
mean   7.383634e+17           6.978112e+17         4.196984e+09   
std    6.780118e+16           4.359384e+16         0.000000e+00   
min    6.660209e+17           6.671522e+17         4.196984e+09   
25%    6.764706e+17           6.732411e+17         4.196984e+09   
50%    7.119681e+17           6.757073e+17         4.196984e+09   
75%    7.931959e+17           7.031489e+17         4.196984e+09   
max    8.924206e+17           8.558181e+17         4.196984e+09   

       retweeted_status_id  retweeted_status_user_id       rating  \
count         7.900000e+01              7.900000e+01  2073.000000   
mean          7.663364e+17              9.967874e+15     1.169987   
std           5.882530e+16              8.859637e+16     3.988880   
min           6.675094e+17              7.832140e+05     0.000000   
25%           7.124037e+17              4.196984e+09     1.000000   
50%           7.798343e+17              4.196984e+09     1.100000   
75%           8.068680e+17              4.196984e+09     1.200000   
max           8.768508e+17              7.874618e+17   177.600000   

       retweet_count  favorite_count      img_num      p1_conf       p2_conf  \
count    2073.000000     2073.000000  2073.000000  2073.000000  2.073000e+03   
mean     2976.089243     8556.718283     1.203570     0.594532  1.346665e-01   
std      5054.897526    12098.640994     0.561856     0.271234  1.006830e-01   
min        16.000000        0.000000     1.000000     0.044333  1.011300e-08   
25%       634.000000     1674.000000     1.000000     0.364095  5.390140e-02   
50%      1408.000000     3864.000000     1.000000     0.588230  1.186220e-01   
75%      3443.000000    10937.000000     1.000000     0.843911  1.955730e-01   
max     79515.000000   132810.000000     4.000000     1.000000  4.880140e-01   

            p3_conf  
count  2.073000e+03  
mean   6.034005e-02  
std    5.092769e-02  
min    1.740170e-10  
25%    1.619920e-02  
50%    4.947150e-02  
75%    9.193000e-02  
max    2.734190e-01

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.073000e+03           2.300000e+01         2.300000e+01   
mean   7.383634e+17           6.978112e+17         4.196984e+09   
std    6.780118e+16           4.359384e+16         0.000000e+00   
min    6.660209e+17           6.671522e+17         4.196984e+09   
25%    6.764706e+17           6.732411e+17         4.196984e+09   
50%    7.119681e+17           6.757073e+17         4.196984e+09   
75%    7.931959e+17           7.031489e+17         4.196984e+09   
max    8.924206e+17           8.558181e+17         4.196984e+09   

       retweeted_status_id  retweeted_status_user_id       rating  \
count         7.900000e+01              7.900000e+01  2073.000000   
mean          7.663364e+17              9.967874e+15     1.169987   
std           5.882530e+16              8.859637e+16     3.988880   
min           6.675094e+17              7.832140e+05     0.000000   
25%           7.124037e+17              4.196984e+09     1.000000   
50%           7.798343e+17              4.196984e+09     1.100000   
75%           8.068680e+17              4.196984e+09     1.200000   
max           8.768508e+17              7.874618e+17   177.600000   

       retweet_count  favorite_count      img_num      p1_conf       p2_conf  \
count    2073.000000     2073.000000  2073.000000  2073.000000  2.073000e+03   
mean     2976.089243     8556.718283     1.203570     0.594532  1.346665e-01   
std      5054.897526    12098.640994     0.561856     0.271234  1.006830e-01   
min        16.000000        0.000000     1.000000     0.044333  1.011300e-08   
25%       634.000000     1674.000000     1.000000     0.364095  5.390140e-02   
50%      1408.000000     3864.000000     1.000000     0.588230  1.186220e-01   
75%      3443.000000    10937.000000     1.000000     0.843911  1.955730e-01   
max     79515.000000   132810.000000     4.000000     1.000000  4.880140e-01   

            p3_conf  
count  2.073000e+03  
mean   6.034005e-02  
std    5.092769e-02  
min    1.740170e-10  
25%    1.619920e-02  
50%    4.947150e-02  
75%    9.193000e-02  
max    2.734190e-01

### Cleaning Quality Issues 1;high amount of missing values in some columns

### Define

#### columns with high amount of missing values will be drop in other not to affect our analysis

### Code

In [28]:
#to check for percentage of missing values in the data
master_archive.isnull().sum()/len(master_archive)*100

tweet_id                       0.000000
in_reply_to_status_id         98.890497
in_reply_to_user_id           98.890497
timestamp                      0.000000
source                         0.000000
text                           0.000000
retweeted_status_id           96.189098
retweeted_status_user_id      96.189098
retweeted_status_timestamp    96.189098
expanded_urls                  0.000000
name                           0.000000
rating                         0.000000
breed_type                    84.563435
retweet_count                  0.000000
favorite_count                 0.000000
jpg_url                        0.000000
img_num                        0.000000
p1                             0.000000
p1_conf                        0.000000
p1_dog                         0.000000
p2                             0.000000
p2_conf                        0.000000
p2_dog                         0.000000
p3                             0.000000
p3_conf                        0.000000


tweet_id                       0.000000
in_reply_to_status_id         98.890497
in_reply_to_user_id           98.890497
timestamp                      0.000000
source                         0.000000
text                           0.000000
retweeted_status_id           96.189098
retweeted_status_user_id      96.189098
retweeted_status_timestamp    96.189098
expanded_urls                  0.000000
name                           0.000000
rating                         0.000000
breed_type                    84.563435
retweet_count                  0.000000
favorite_count                 0.000000
jpg_url                        0.000000
img_num                        0.000000
p1                             0.000000
p1_conf                        0.000000
p1_dog                         0.000000
p2                             0.000000
p2_conf                        0.000000
p2_dog                         0.000000
p3                             0.000000
p3_conf                        0.000000


In [29]:
# dropping the columns with null value percentage higher than 90%
master_archive.drop(columns=['in_reply_to_status_id','in_reply_to_user_id','retweeted_status_id','retweeted_status_user_id','retweeted_status_timestamp'], inplace = True)

### Test


In [30]:
# checking the info after the columns has being dropped
master_archive.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2073 entries, 0 to 2072
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   tweet_id        2073 non-null   int64  
 1   timestamp       2073 non-null   object 
 2   source          2073 non-null   object 
 3   text            2073 non-null   object 
 4   expanded_urls   2073 non-null   object 
 5   name            2073 non-null   object 
 6   rating          2073 non-null   float64
 7   breed_type      320 non-null    object 
 8   retweet_count   2073 non-null   int64  
 9   favorite_count  2073 non-null   int64  
 10  jpg_url         2073 non-null   object 
 11  img_num         2073 non-null   int64  
 12  p1              2073 non-null   object 
 13  p1_conf         2073 non-null   float64
 14  p1_dog          2073 non-null   bool   
 15  p2              2073 non-null   object 
 16  p2_conf         2073 non-null   float64
 17  p2_dog          2073 non-null   b


### Cleaning Quality Issues 2;expanded_url column not really important because it does not contain the image

### Define 

#### expanded_urls column will be dropped

### Code

In [31]:
#  drop the expanded_urlcolumn
master_archive.drop(columns='expanded_urls',inplace = True)

### Test

In [32]:
master_archive.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2073 entries, 0 to 2072
Data columns (total 20 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   tweet_id        2073 non-null   int64  
 1   timestamp       2073 non-null   object 
 2   source          2073 non-null   object 
 3   text            2073 non-null   object 
 4   name            2073 non-null   object 
 5   rating          2073 non-null   float64
 6   breed_type      320 non-null    object 
 7   retweet_count   2073 non-null   int64  
 8   favorite_count  2073 non-null   int64  
 9   jpg_url         2073 non-null   object 
 10  img_num         2073 non-null   int64  
 11  p1              2073 non-null   object 
 12  p1_conf         2073 non-null   float64
 13  p1_dog          2073 non-null   bool   
 14  p2              2073 non-null   object 
 15  p2_conf         2073 non-null   float64
 16  p2_dog          2073 non-null   bool   
 17  p3              2073 non-null   o

### Cleaning Quality Issues 3; The empty values stored as none in name column should be replaced with NAN

 ### Define

#### Entries stored as None will be replaced with NAN

### Code

In [33]:
master_archive['name'] =master_archive['name'].replace('None', np.nan)

### Test

In [34]:
master_archive.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2073 entries, 0 to 2072
Data columns (total 20 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   tweet_id        2073 non-null   int64  
 1   timestamp       2073 non-null   object 
 2   source          2073 non-null   object 
 3   text            2073 non-null   object 
 4   name            1496 non-null   object 
 5   rating          2073 non-null   float64
 6   breed_type      320 non-null    object 
 7   retweet_count   2073 non-null   int64  
 8   favorite_count  2073 non-null   int64  
 9   jpg_url         2073 non-null   object 
 10  img_num         2073 non-null   int64  
 11  p1              2073 non-null   object 
 12  p1_conf         2073 non-null   float64
 13  p1_dog          2073 non-null   bool   
 14  p2              2073 non-null   object 
 15  p2_conf         2073 non-null   float64
 16  p2_dog          2073 non-null   bool   
 17  p3              2073 non-null   o

### Cleaning Quality Issues 4;timestamp datatype is stored as object instead of datetime

### Define

Timestamp data will be changed fom object to datetime

### Code

In [35]:
# changing timestamp datatype which is object to datetime
master_archive["timestamp"] =master_archive["timestamp"].astype('datetime64[ns]')

### Test

In [36]:
# check of the datatype has effectively changed
master_archive.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2073 entries, 0 to 2072
Data columns (total 20 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   tweet_id        2073 non-null   int64         
 1   timestamp       2073 non-null   datetime64[ns]
 2   source          2073 non-null   object        
 3   text            2073 non-null   object        
 4   name            1496 non-null   object        
 5   rating          2073 non-null   float64       
 6   breed_type      320 non-null    object        
 7   retweet_count   2073 non-null   int64         
 8   favorite_count  2073 non-null   int64         
 9   jpg_url         2073 non-null   object        
 10  img_num         2073 non-null   int64         
 11  p1              2073 non-null   object        
 12  p1_conf         2073 non-null   float64       
 13  p1_dog          2073 non-null   bool          
 14  p2              2073 non-null   object        
 15  p2_c

 ###  Cleaning Quality Issues 5 ; in p1,p2,p3 columns,some description are with underscore while others are not

### Define

#### The underscore will be removed so all description will be in same format

### Code

In [37]:
# replace the undescore with space
master_archive['p1'] = master_archive['p1'].str.replace('_', ' ')
master_archive['p2'] = master_archive['p2'].str.replace('_', ' ')
master_archive['p3'] = master_archive['p3'].str.replace('_', ' ')

### Test

In [38]:
master_archive.head()

tweet_id           timestamp  \
0  892420643555336193 2017-08-01 16:23:56   
1  892177421306343426 2017-08-01 00:17:27   
2  891815181378084864 2017-07-31 00:18:03   
3  891689557279858688 2017-07-30 15:58:51   
4  891327558926688256 2017-07-29 16:00:24   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text      name  rating  \
0  This is Phineas. He's a mystical boy. Only eve...   Phineas     1.3   
1  This is Tilly. She's just checking pup on you....     Tilly     1.3   
2  This is Archie. He is a rare Norwegian Pouncin...    Archie     1.2   
3  This is Darla. She commenced a snooze mid meal...     Darla     1.3   
4  This is Franklin. He would like you to stop ca...  Franklin     1.2   

  breed_type  retweet_count  favorite_count  \
0        NaN           8853           39467   
1        NaN           6514           33819   
2        NaN           4328           25461   
3        NaN           8964           42908   
4        NaN           9774           41048   

                                           jpg_url  img_num           p1  \
0  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg        1       orange   
1  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg        1    Chihuahua   
2  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg        1    Chihuahua   
3  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg        1  paper towel   
4  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg        2       basset   

    p1_conf  p1_dog                  p2   p2_conf  p2_dog  \
0  0.097049   False               bagel  0.085851   False   
1  0.323581    True            Pekinese  0.090647    True   
2  0.716012    True            malamute  0.078253    True   
3  0.170278   False  Labrador retriever  0.168086    True   
4  0.555712    True    English springer  0.225770    True   

                            p3   p3_conf  p3_dog  
0                       banana  0.076110   False  
1                     papillon  0.068957    True  
2                       kelpie  0.031379    True  
3                      spatula  0.040836   False  
4  German short-haired pointer  0.175219    True

tweet_id           timestamp  \
0  892420643555336193 2017-08-01 16:23:56   
1  892177421306343426 2017-08-01 00:17:27   
2  891815181378084864 2017-07-31 00:18:03   
3  891689557279858688 2017-07-30 15:58:51   
4  891327558926688256 2017-07-29 16:00:24   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text      name  rating  \
0  This is Phineas. He's a mystical boy. Only eve...   Phineas     1.3   
1  This is Tilly. She's just checking pup on you....     Tilly     1.3   
2  This is Archie. He is a rare Norwegian Pouncin...    Archie     1.2   
3  This is Darla. She commenced a snooze mid meal...     Darla     1.3   
4  This is Franklin. He would like you to stop ca...  Franklin     1.2   

  breed_type  retweet_count  favorite_count  \
0        NaN           8853           39467   
1        NaN           6514           33819   
2        NaN           4328           25461   
3        NaN           8964           42908   
4        NaN           9774           41048   

                                           jpg_url  img_num           p1  \
0  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg        1       orange   
1  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg        1    Chihuahua   
2  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg        1    Chihuahua   
3  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg        1  paper towel   
4  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg        2       basset   

    p1_conf  p1_dog                  p2   p2_conf  p2_dog  \
0  0.097049   False               bagel  0.085851   False   
1  0.323581    True            Pekinese  0.090647    True   
2  0.716012    True            malamute  0.078253    True   
3  0.170278   False  Labrador retriever  0.168086    True   
4  0.555712    True    English springer  0.225770    True   

                            p3   p3_conf  p3_dog  
0                       banana  0.076110   False  
1                     papillon  0.068957    True  
2                       kelpie  0.031379    True  
3                      spatula  0.040836   False  
4  German short-haired pointer  0.175219    True

### Cleaning Quality Issues 6 ; some Column headers are not discriptive enough

### Define 


#### The less descriptive column will be modify to make more sense

### Code

In [39]:
# renaming the affected columns
master_archive = master_archive.rename(columns = {'img_num':'image_number',
                                                  'p1': 'prediction_1',
                                                  'p1_conf':'prediction_1_confidence',
                                                  'p1_dog':'prediction_1_dog',
                                                  'p2': 'prediction_2',
                                                  'p2_conf':'prediction_2_confidence',
                                                  'p2_dog':'prediction_2_dog',
                                                  'p3': 'prediction_3',
                                                  'p3_conf':'prediction_3_confidence',
                                                  'p3_dog':'prediction_3_dog'})

### Test

In [40]:
# to check if the column headers have change
master_archive.head(2)

tweet_id           timestamp  \
0  892420643555336193 2017-08-01 16:23:56   
1  892177421306343426 2017-08-01 00:17:27   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   

                                                text     name  rating  \
0  This is Phineas. He's a mystical boy. Only eve...  Phineas     1.3   
1  This is Tilly. She's just checking pup on you....    Tilly     1.3   

  breed_type  retweet_count  favorite_count  \
0        NaN           8853           39467   
1        NaN           6514           33819   

                                           jpg_url  image_number prediction_1  \
0  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg             1       orange   
1  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg             1    Chihuahua   

   prediction_1_confidence  prediction_1_dog prediction_2  \
0                 0.097049             False        bagel   
1                 0.323581              True     Pekinese   

   prediction_2_confidence  prediction_2_dog prediction_3  \
0                 0.085851             False       banana   
1                 0.090647              True     papillon   

   prediction_3_confidence  prediction_3_dog  
0                 0.076110             False  
1                 0.068957              True

tweet_id           timestamp  \
0  892420643555336193 2017-08-01 16:23:56   
1  892177421306343426 2017-08-01 00:17:27   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   

                                                text     name  rating  \
0  This is Phineas. He's a mystical boy. Only eve...  Phineas     1.3   
1  This is Tilly. She's just checking pup on you....    Tilly     1.3   

  breed_type  retweet_count  favorite_count  \
0        NaN           8853           39467   
1        NaN           6514           33819   

                                           jpg_url  image_number prediction_1  \
0  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg             1       orange   
1  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg             1    Chihuahua   

   prediction_1_confidence  prediction_1_dog prediction_2  \
0                 0.097049             False        bagel   
1                 0.323581              True     Pekinese   

   prediction_2_confidence  prediction_2_dog prediction_3  \
0                 0.085851             False       banana   
1                 0.090647              True     papillon   

   prediction_3_confidence  prediction_3_dog  
0                 0.076110             False  
1                 0.068957              True

### Cleaning Quality Issues 7 ; under the name column,some values appear in lower case and it seems not to be real name

### Define


#### Any Name in lower case will be changed to NAN


### Code

In [41]:
# this output subset of names that are in lowercase
lower_names = master_archive[master_archive['name'].str.islower() == True] 
lower_names

tweet_id           timestamp  \
21    887517139158093824 2017-07-19 03:39:09   
50    881536004380872706 2017-07-02 15:32:16   
141   859196978902773760 2017-05-02 00:04:57   
155   855459453768019968 2017-04-21 16:33:22   
293   828650029636317184 2017-02-06 17:02:17   
...                  ...                 ...   
2066  666051853826850816 2015-11-16 00:35:11   
2067  666050758794694657 2015-11-16 00:30:50   
2069  666044226329800704 2015-11-16 00:04:52   
2070  666033412701032449 2015-11-15 23:21:54   
2071  666029285002620928 2015-11-15 23:05:30   

                                                 source  \
21    <a href="http://twitter.com/download/iphone" r...   
50    <a href="http://twitter.com/download/iphone" r...   
141   <a href="http://twitter.com/download/iphone" r...   
155   <a href="http://twitter.com/download/iphone" r...   
293   <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2066  <a href="http://twitter.com/download/iphone" r...   
2067  <a href="http://twitter.com/download/iphone" r...   
2069  <a href="http://twitter.com/download/iphone" r...   
2070  <a href="http://twitter.com/download/iphone" r...   
2071  <a href="http://twitter.com/download/iphone" r...   

                                                   text   name  rating  \
21    I've yet to rate a Venezuelan Hover Wiener. Th...   such     1.4   
50    Here is a pupper approaching maximum borkdrive...      a     1.4   
141   We only rate dogs. This is quite clearly a smo...  quite     1.2   
155   Guys, we only rate dogs. This is quite clearly...  quite     1.2   
293   Occasionally, we're sent fantastic stories. Th...    one     1.4   
...                                                 ...    ...     ...   
2066  This is an odd dog. Hard on the outside but lo...     an     0.2   
2067  This is a truly beautiful English Wilson Staff...      a     1.0   
2069  This is a purebred Piers Morgan. Loves to Netf...      a     0.6   
2070  Here is a very happy pup. Big fan of well-main...      a     0.9   
2071  This is a western brown Mitsubishi terrier. Up...      a     0.7   

     breed_type  retweet_count  favorite_count  \
21          NaN          12082           46959   
50       pupper          16570           50199   
141         NaN          25661           75193   
155         NaN           8987           31657   
293         NaN           1544           10467   
...         ...            ...             ...   
2066        NaN            879            1253   
2067        NaN             60             136   
2069        NaN            147             311   
2070        NaN             47             128   
2071        NaN             48             132   

                                                jpg_url  image_number  \
21    https://pbs.twimg.com/ext_tw_video_thumb/88751...             1   
50    https://pbs.twimg.com/ext_tw_video_thumb/88153...             1   
141   https://pbs.twimg.com/ext_tw_video_thumb/85919...             1   
155     https://pbs.twimg.com/media/C98z1ZAXsAEIFFn.jpg             2   
293     https://pbs.twimg.com/media/C3_0yhCWEAETXj2.jpg             1   
...                                                 ...           ...   
2066    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg             1   
2067    https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg             1   
2069    https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg             1   
2070    https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg             1   
2071    https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg             1   

              prediction_1  prediction_1_confidence  prediction_1_dog  \
21               limousine                 0.130432             False   
50                 Samoyed                 0.281463              True   
141                 Angora                 0.224218             False   
155       Blenheim spaniel                 0.389513              True  

tweet_id           timestamp  \
21    887517139158093824 2017-07-19 03:39:09   
50    881536004380872706 2017-07-02 15:32:16   
141   859196978902773760 2017-05-02 00:04:57   
155   855459453768019968 2017-04-21 16:33:22   
293   828650029636317184 2017-02-06 17:02:17   
...                  ...                 ...   
2066  666051853826850816 2015-11-16 00:35:11   
2067  666050758794694657 2015-11-16 00:30:50   
2069  666044226329800704 2015-11-16 00:04:52   
2070  666033412701032449 2015-11-15 23:21:54   
2071  666029285002620928 2015-11-15 23:05:30   

                                                 source  \
21    <a href="http://twitter.com/download/iphone" r...   
50    <a href="http://twitter.com/download/iphone" r...   
141   <a href="http://twitter.com/download/iphone" r...   
155   <a href="http://twitter.com/download/iphone" r...   
293   <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2066  <a href="http://twitter.com/download/iphone" r...   
2067  <a href="http://twitter.com/download/iphone" r...   
2069  <a href="http://twitter.com/download/iphone" r...   
2070  <a href="http://twitter.com/download/iphone" r...   
2071  <a href="http://twitter.com/download/iphone" r...   

                                                   text   name  rating  \
21    I've yet to rate a Venezuelan Hover Wiener. Th...   such     1.4   
50    Here is a pupper approaching maximum borkdrive...      a     1.4   
141   We only rate dogs. This is quite clearly a smo...  quite     1.2   
155   Guys, we only rate dogs. This is quite clearly...  quite     1.2   
293   Occasionally, we're sent fantastic stories. Th...    one     1.4   
...                                                 ...    ...     ...   
2066  This is an odd dog. Hard on the outside but lo...     an     0.2   
2067  This is a truly beautiful English Wilson Staff...      a     1.0   
2069  This is a purebred Piers Morgan. Loves to Netf...      a     0.6   
2070  Here is a very happy pup. Big fan of well-main...      a     0.9   
2071  This is a western brown Mitsubishi terrier. Up...      a     0.7   

     breed_type  retweet_count  favorite_count  \
21          NaN          12082           46959   
50       pupper          16570           50199   
141         NaN          25661           75193   
155         NaN           8987           31657   
293         NaN           1544           10467   
...         ...            ...             ...   
2066        NaN            879            1253   
2067        NaN             60             136   
2069        NaN            147             311   
2070        NaN             47             128   
2071        NaN             48             132   

                                                jpg_url  image_number  \
21    https://pbs.twimg.com/ext_tw_video_thumb/88751...             1   
50    https://pbs.twimg.com/ext_tw_video_thumb/88153...             1   
141   https://pbs.twimg.com/ext_tw_video_thumb/85919...             1   
155     https://pbs.twimg.com/media/C98z1ZAXsAEIFFn.jpg             2   
293     https://pbs.twimg.com/media/C3_0yhCWEAETXj2.jpg             1   
...                                                 ...           ...   
2066    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg             1   
2067    https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg             1   
2069    https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg             1   
2070    https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg             1   
2071    https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg             1   

              prediction_1  prediction_1_confidence  prediction_1_dog  \
21               limousine                 0.130432             False   
50                 Samoyed                 0.281463              True   
141                 Angora                 0.224218             False   
155       Blenheim spaniel                 0.389513              True  

In [42]:
# changing the names in lower case to NAN
master_archive[master_archive['name'].str.islower() == True]='NAN'

### Test

In [43]:
# recheck if there is still any name in lowercase
master_archive[master_archive['name'].str.islower() == True] 

Empty DataFrame
Columns: [tweet_id, timestamp, source, text, name, rating, breed_type, retweet_count, favorite_count, jpg_url, image_number, prediction_1, prediction_1_confidence, prediction_1_dog, prediction_2, prediction_2_confidence, prediction_2_dog, prediction_3, prediction_3_confidence, prediction_3_dog]
Index: []

Empty DataFrame
Columns: [tweet_id, timestamp, source, text, name, rating, breed_type, retweet_count, favorite_count, jpg_url, image_number, prediction_1, prediction_1_confidence, prediction_1_dog, prediction_2, prediction_2_confidence, prediction_2_dog, prediction_3, prediction_3_confidence, prediction_3_dog]
Index: []

### Cleaning Quality Issues 8; presence ofoutlier in rating_proportion,retweet_count and varorite favorite_count

### Define

#### The outlier should be removed, but i dont have understanding of how to do that at this stage

In [44]:
# saving th clean dataframe in a csv file
master_archive.to_csv('twitter_archive_master.csv', index=False)

In [45]:
y1=pd.read_csv('twitter_archive_master.csv')

In [46]:
y1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2073 entries, 0 to 2072
Data columns (total 20 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   tweet_id                 2073 non-null   object
 1   timestamp                2073 non-null   object
 2   source                   2073 non-null   object
 3   text                     2073 non-null   object
 4   name                     1496 non-null   object
 5   rating                   2073 non-null   object
 6   breed_type               407 non-null    object
 7   retweet_count            2073 non-null   object
 8   favorite_count           2073 non-null   object
 9   jpg_url                  2073 non-null   object
 10  image_number             2073 non-null   object
 11  prediction_1             2073 non-null   object
 12  prediction_1_confidence  2073 non-null   object
 13  prediction_1_dog         2073 non-null   object
 14  prediction_2             2073 non-null  

In [47]:
y1.dropna(subset =['name'],inplace=True)

In [48]:
y1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1496 entries, 0 to 2071
Data columns (total 20 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   tweet_id                 1496 non-null   object
 1   timestamp                1496 non-null   object
 2   source                   1496 non-null   object
 3   text                     1496 non-null   object
 4   name                     1496 non-null   object
 5   rating                   1496 non-null   object
 6   breed_type               284 non-null    object
 7   retweet_count            1496 non-null   object
 8   favorite_count           1496 non-null   object
 9   jpg_url                  1496 non-null   object
 10  image_number             1496 non-null   object
 11  prediction_1             1496 non-null   object
 12  prediction_1_confidence  1496 non-null   object
 13  prediction_1_dog         1496 non-null   object
 14  prediction_2             1496 non-null  